### ISSUE TITLE
Integration with pandas

### ISSUE THREAD

**Issue by yiga2** [fastavro/fastavro#112]
Now that fasparquet is integrated with pandas, any talk/plan to extend to fastavro ?

**Comment by tebeka** [comment]
We don't have plan for that. You mean to have `DataFrame.to_avro`? That's up to the Pandas team, we'll gladly help.

**Comment by barrywhart** [comment]
@yiga2: Can you provide any details, documentation, examples, etc. on the fastparquet-pandas integration?

This issue may overlap somewhat with #76, which requests allowing fastavro's writer to accept records in list or tuple/namedtuple format (in addition to the currently supported dictionary format).

**Comment by barrywhart** [comment]
Section 5.2.1 of [this PDF](https://media.readthedocs.org/pdf/fastparquet/latest/fastparquet.pdf) has some documentation. If I read this correctly,  it seems that the integration is in the fastparquet library, not the pandas library.

**Comment by barrywhart** [comment]
Source code for `to_pandas()` is [here](https://github.com/dask/fastparquet/blob/master/fastparquet/api.py#L350).

**Comment by barrywhart** [comment]
Both fastparquet and the Apache Avro project from which fastavro was forked both appear to use the same Apache 2.0 license, so it seems like fastavro could reuse code from fastparquet's Pandas integration, if it seemed useful to do so.

**Comment by NightFantomJ2** [comment]
Both numpy's integral types (np.int64 etc) and python's give true on `isinstance(num, numbers.Integral)`
[https://docs.python.org/3/library/numbers.html](https://docs.python.org/3/library/numbers.html), 

changing the isinstance lines for int and long here:
https://github.com/tebeka/fastavro/blob/8b2d2e533da9d890978380f83a04eaa41d6a7527/fastavro/_write_py.py#L330-L344
and here:
https://github.com/tebeka/fastavro/blob/8b2d2e533da9d890978380f83a04eaa41d6a7527/fastavro/_write.pyx#L429-L443

should fix all integral numpy integration.
(works locally without any other changes)
Analogous fixes could be done for floating point types and so forth I presume.

Any reasons not to? I can write up a pull request if you want.

**Comment by scottbelden** [comment]
I'll be honest; I don't use `pandas` at all and `numpy` all that much, so I've never really understood what the goal of this issue is. However, if it can be resolved by making the change to `numbers.Integral` that you mentioned, then that sounds good to me.

Feel free to submit the PR. Thanks!

**Comment by barrywhart** [comment]
Seems reasonable to me. Can you include tests for the new data types being supported (i.e. the numpy types)?

@scottbelden: This sounds like a useful change, but I don't think it will resolve the overall issue. I think the overall request is to convert Avro records into a Pandas DataFrame, an in-memory table format with some SQL-like capabilities. Data scientists love it for loading up data, looking at it, manipulating it, etc.

**Comment by rhaarm** [comment]
I've been planning on working on the pandas integration `to/from_avro()` or `to/from_pandas()` on both projects. I'm just planning to do it after the custom types work. Also, possibly do some work on the Pandas project for complete integration like Parquet.

**Comment by NightFantomJ2** [comment]
After some more looking into it, I concur that it won't solve the entire pandas DataFrame integration, but this is due to pandas converting a Series (=column) of type `['null', 'long']` to a (numpy) float. This is because in numpy (used underlying in pandas) a list needs to contain a uniform type (for efficiency reasons), and a float can describe both NaN and longs. 

In other words an avro column with type ['null', 'long'] containing [0, 2, 4, NULL, 8] would be read into pandas as [0.0, 2.0, 4.0, NaN, 8.0], and then written back as ['null', 'float'] if the schema is inferred from the column types in pandas.

As-is this isn't a big issue (should be a lossless type conversion) but it's something to keep in mind, and might be less obvious to numpy/pandas users than just receiving a type error.

Sidenote, when running pytest with no arguments in the root dir I get this error:
`AttributeError: module 'fastavro._read_py' has no attribute 'CYTHON_MODULE'`
I've never worked with cython before, but I've installed it through anaconda, isn't that enough? Or is it a different error?

**Comment by barrywhart** [comment]
This sounds great! I am starting to use Pandas on some projects now, and this will be a nice integration. The specific change on numpy type handling as well as the overall Pandas integration sound like good directions to me.

**Comment by kkirsanov** [comment]
I recently came across a project - https://github.com/ynqa/pandavro

**Comment by IanQS** [comment]
Any updates on this? Would love to see what comes of this

**Comment by scottbelden** [comment]
@IanQS I don't think it was ever clear what exactly this means or examples of how it would work. I don't use `pandas` very much at all so it's hard for me to judge what the level of effort is for this or how feasible it is.

**Comment by IanQS** [comment]
@scottbelden Totally understood! I suppose that comment was directed more at @JoranDox but I suppose the right place to talk about this would be on the Pandas end. I stumbled on this when I googled pandas fastavro

**Comment by tswast** [comment]
I did a little experimentation on this a while back and determined that jitted code is really important for the pandas use-case, because we need to go from row-oriented to column-oriented. This can involve a lot of extra copies & passes over the data.

I wrote  a prototype Avro-to-Arrow parser to demonstrate some performance improvement possibilities. (The reason being that Arrow has a standard in-memory format for more data types and has fast/no-copy conversions to Pandas)

Code: https://github.com/tswast/code-snippets/tree/master/2019/avro-to-arrow
Explanation as a lightning talk at SciPy 2019: https://youtu.be/AnYjArI2xUM?t=720

There was an Apache Arrow C++ PR to implement some of these ideas in https://github.com/apache/arrow/pull/4585 but that work has stalled.
